contracts/schemes/ArcUpgradeScheme.sol

```diff
@@ -0,0 +1,200 @@
+    function executeProposal(bytes32 _proposalId, int256 _decision)
+    external
+    onlyVotingMachine(_proposalId)
+    returns(bool) {
+        UpgradeProposal storage proposal = organizationProposals[_proposalId];
+        require(proposal.exist, "must be a live proposal");
+        require(proposal.passed == false, "cannot execute twice");
+
+        if (_decision == 1) {
+            proposal.passed = true;
+            execute(_proposalId);
+        } else {
+            delete organizationProposals[_proposalId];
+            emit ProposalDeleted(address(avatar), _proposalId);
+        }
+
+        emit ProposalExecutedByVotingMachine(address(avatar), _proposalId, _decision);
+        return true;
+    }
+
+    /**
+    * @dev execution of proposals after it has been decided by the voting machine
+    * @param _proposalId the ID of the voting in the voting machine
+    */
+    function execute(bytes32 _proposalId) public {
+        UpgradeProposal storage proposal = organizationProposals[_proposalId];
+        require(proposal.exist, "must be a live proposal");
+        require(proposal.passed, "proposal must passed by voting machine");
+        proposal.exist = false;
+        address[] memory contractsToUpgrade = proposal.contractsToUpgrade;
+        for (uint256 i = 0; i < contractsToUpgrade.length; i++) {
+            bytes32 contractNameBytes = proposal.contractsNames[i];
+            string memory contractName = bytes32ToStr(contractNameBytes);
+            address updatedImp = ImplementationProvider(
+                package.getContract(proposal.packageVersion)
+            ).getImplementation(contractName);
+
+            Controller controller = Controller(avatar.owner());
+            controller.genericCall(
+                contractsToUpgrade[i],
+                abi.encodeWithSignature("upgradeTo(address)", updatedImp),
+                0
+            );
+        }
+
+        delete organizationProposals[_proposalId];
+        emit ProposalDeleted(address(avatar), _proposalId);
+        emit ProposalExecuted(address(avatar), _proposalId);
+    }
+
+    /**
+    * @dev propose upgrade contracts Arc version
+    *      The function trigger NewUpgradeProposal event
+    * @param _packageVersion - the new Arc version to use for the contracts
+    * @param _contractsNames - names of contracts which needs to be upgraded
+    * @param _contractsToUpgrade - addresses of contracts which needs to be upgraded
+    * @param _descriptionHash - proposal description hash
+    * @return an id which represents the proposal
+    */
+    function proposeUpgrade(
+        uint64[3] memory _packageVersion,
+        bytes32[] memory _contractsNames,
+        address[] memory _contractsToUpgrade,
+        string memory _descriptionHash)
+    public
+    returns(bytes32)
+    {
+        require(package.hasVersion(_packageVersion), "Specified version doesn't exist in the Package");
+        for (uint256 i = 0; i < _contractsToUpgrade.length; i++) {
```

Comment:
I think we should get the max possible length which can be executed gas wised and cap the length of this array with that

Reply:
20 seems to be around the max, so will cap it there ok?

Reply:
does the number 60 verified/tested ?

Reply:
It is tested that 60 works. I think more is possible theoretically but not much so just rounded it down.